packages/vite/src/node/config.ts

```diff
@@ -522,35 +541,56 @@ export async function resolveConfig(
     spa: config.spa ?? true
   }
 
-  // Some plugins that aren't intended to work in the bundling of workers (doing post-processing at build time for example).
-  // And Plugins may also have cached that could be corrupted by being used in these extra rollup calls.
-  // So we need to separate the worker plugin from the plugin that vite needs to run.
-  const [workerPrePlugins, workerNormalPlugins, workerPostPlugins] =
-    sortUserPlugins(config.worker?.plugins as Plugin[])
   const workerResolved: ResolvedConfig = {
     ...resolved,
     isWorker: true,
     mainConfig: resolved
   }
-  resolved.worker.plugins = await resolvePlugins(
-    workerResolved,
-    workerPrePlugins,
-    workerNormalPlugins,
-    workerPostPlugins
-  )
-  // call configResolved worker plugins hooks
-  await Promise.all(
-    resolved.worker.plugins.map((p) => p.configResolved?.(workerResolved))
-  )
+
   ;(resolved.plugins as Plugin[]) = await resolvePlugins(
     resolved,
     prePlugins,
     normalPlugins,
     postPlugins
   )
 
+  // Some plugins that aren't intended to work in the bundling of workers (doing post-processing at build time for example).
+  // And Plugins may also have cached that could be corrupted by being used in these extra rollup calls.
+  // So we need to separate the worker plugin from the plugin that vite needs to run.
+  // serve mode doesn't bundle the worker so doesn't resolve plugin again
+  let userWorkerPlugins: Plugin[] = []
+  if (command === 'build') {
+    const [workerPrePlugins, workerNormalPlugins, workerPostPlugins] =
+      sortUserPlugins(rawWorkerUserPlugins)
+
+    // run config hooks
+    userWorkerPlugins = [
+      ...workerPrePlugins,
+      ...workerNormalPlugins,
+      ...workerPostPlugins
+    ]
+    for (const p of userWorkerPlugins) {
+      if (p.config) {
+        const res = await p.config(config, configEnv)
+        if (res) {
+          config = mergeConfig(config, res)
+        }
+      }
+    }
```

Comment:
Splitting out the config hook here could cause inconsistency with dev & build since:

in dev: normal + worker are merged and applies to `config` hook
in build: normal applies `config` hook, and then worker applies `config` hook.

It changes depending on the plugins `enforce` prop 🤔